### PR TITLE
Add Hdbscan trait, public API, and benchmark

### PR DESCRIPTION
**Pull request by urschrei**
Completes the HDBSCAN stack: the `Hdbscan` trait (with an impl for `MultiPoint` and slices of `Point`), the `HdbscanResult` and `HdbscanError` types, input validation, and the `hdbscan_impl` orchestration that wires the pipeline steps together. Exported from the algorithm module alongside `Dbscan`, with the crate docs updated.

Removes the temporary `#![allow(dead_code)]` now that every internal item has a caller.

Adds the known-answer tests (labels and GLOSH scores validated against the reference python `hdbscan` library), end-to-end and edge case tests, and a criterion benchmark using the shared `make_moons` generator.

---

Part of the HDBSCAN stack (6 of 6). Based on #1557.

### CODE REVIEW

File: geo/src/algorithm/hdbscan/mod.rs [hunk 1]
```diff
@@ -10,19 +10,397 @@
...
+
+    /// GLOSH (Global-Local Outlier Scores from Hierarchies) outlier score for
+    /// each input point, in the range [0, 1]. Higher values indicate stronger
+    /// outliers.
+    pub outlier_scores: Vec<T>,
+}
+
+/// Errors that can occur during [`hdbscan`](Hdbscan::hdbscan) clustering.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum HdbscanError {
```
Comment (michaelkirk) [inline geo/src/algorithm/hdbscan/mod.rs:40]: Please use `thiserror`

File: geo/src/algorithm/hdbscan/mod.rs [hunk 2]
```diff
@@ -10,19 +10,397 @@
...
+/// // Points in cluster 1 share the same label
+/// assert_eq!(result.labels[0], result.labels[1]);
+/// // Points in cluster 2 share a different label
+/// assert_eq!(result.labels[6], result.labels[7]);
+/// // The two clusters have different labels
+/// assert_ne!(result.labels[0], result.labels[6]);
+/// ```
+pub trait Hdbscan<T>
+where
+    T: GeoFloat + Send + Sync,
```
Comment (michaelkirk) [inline geo/src/algorithm/hdbscan/mod.rs:256]: Should we just make GeoNum Send/Sync?

File: geo/src/algorithm/hdbscan/mod.rs [hunk 3]
```diff
@@ -10,19 +10,397 @@
...
+/// [`HdbscanError::InvalidMinSamples`] when `min_samples` is less than 1 or
+/// greater than `n - 1` (the core distance is the distance to the
+/// `min_samples`-th nearest *other* point); and
+/// [`HdbscanError::DisconnectedSpanningTree`] when the input contains
+/// non-finite or extremely large coordinates that leave the
+/// mutual-reachability graph disconnected.
+///
+/// # Trait bounds
+///
+/// This trait requires `T: GeoFloat + Send + Sync`, whereas the sibling
```
Comment (michaelkirk) [inline geo/src/algorithm/hdbscan/mod.rs:174]: This documentation seems a bit overly exhaustive at the expense of clarity for something which affects approximately nobody.

File: geo/src/algorithm/hdbscan/mod.rs [hunk 4]
```diff
@@ -10,19 +10,397 @@
...
+/// the feature) keeps the feature additive. In practice `GeoFloat` is only
+/// implemented for `f32` and `f64`, both of which are `Send + Sync`, so this
+/// costs concrete callers nothing; only code generic over `T: GeoFloat` must
+/// add `+ Send + Sync` to call [`hdbscan`](Hdbscan::hdbscan).
+///
+/// # Algorithm
+///
+/// 1. **Compute core distances** – each point's distance to its
+///    `min_samples`-th nearest neighbour – using a
+///    [ball tree](https://en.wikipedia.org/wiki/Ball_tree) rather than
```
Comment (michaelkirk) [inline geo/src/algorithm/hdbscan/mod.rs:188]: link to our impl?

File: geo/src/algorithm/hdbscan/mod.rs [hunk 5]
```diff
@@ -10,19 +10,397 @@
...
 use std::collections::HashSet;
 
+/// Result of an HDBSCAN clustering run.
+///
+/// Contains both cluster labels and GLOSH outlier scores for every input point.
+#[derive(Debug, Clone, PartialEq)]
+pub struct HdbscanResult<T> {
+    /// Cluster assignment for each input point, in input order.
+    /// `Some(id)` for clustered points, `None` for noise.
+    pub labels: Vec<Option<usize>>,
```
Comment (michaelkirk) [inline geo/src/algorithm/hdbscan/mod.rs]: I wonder if there'd be demand for an inverted form of this method - a way to iterate over the clusters.

Not necessary for this first pass, but I guess think about it as you (presumably?) use this code.

File: geo/src/algorithm/hdbscan/mod.rs [hunk 6]
```diff
@@ -10,19 +10,397 @@
...
+///  d ───┴─────┘       ├───    one; the isolated point e is absorbed
+///  e ─────────────────┘       last, at a much larger distance
+/// ```
+///
+/// This is a *single-linkage* hierarchy: the distance between two groups is
+/// simply the distance between their two closest members. Clusters that
+/// persist over a wide range of thresholds (like `{a, b}`) are "stable";
+/// short-lived groupings are artefacts of a particular threshold. HDBSCAN
+/// condenses the dendrogram and keeps only the most stable clusters, and
+/// points such as `e` that never joined a stable cluster become noise.
```
Comment (michaelkirk) [inline geo/src/algorithm/hdbscan/mod.rs:142]: Great docs!

File: geo/src/algorithm/hdbscan/tests.rs [hunk 7]
```diff
@@ -1,7 +1,296 @@
 use super::*;
+use crate::point;
 use crate::wkt;
 use approx::assert_relative_eq;
 
+#[test]
+fn test_empty_input() {
+    let points: Vec<Point<f64>> = vec![];
+    let result = points.hdbscan(5).unwrap();
```
Comment (michaelkirk) [inline geo/src/algorithm/hdbscan/tests.rs:9]: It's admittedly subjective, but I'm initially surprised this does not error while the `test_single_point` test returns an error.

My first instinct is that either `test_empty_input` should return:
```
Err(HdbscanError::InvalidMinClusterSize {
            min_cluster_size: 5,
            n: 0,
        }),
```

Or that `test_single_point` should simply return a valid result with no points being assigned to a cluster since none meet the criteria.

I realize you have this all spelled out in test_empty_vs_tiny_input_asymmetry - it's intentionally not symmetrical, but it didn't align with my guess as to what the behavior would be. If this is consistent with some existing reference implementation or how you think people will actually want to use it, feel free to leave it as is.

File: geo/src/algorithm/hdbscan/tests.rs [hunk 8]
```diff
@@ -1,7 +1,296 @@
...
+            n: 3,
+        }),
+    );
+}
+
+/// Verify that the MultiPoint trait impl delegates correctly and
+/// produces the same result as calling on the underlying slice.
+#[test]
+fn test_multipoint_trait_impl() {
+    let mp = MultiPoint::new(vec![
```
Comment (michaelkirk) [inline geo/src/algorithm/hdbscan/tests.rs:79]: Can you please ask your llm to use `wkt!(MULTIPOINT ...)`  for test fixtures so I can easily visualize these ( and for the `[]` slice variants)

File: geo/src/algorithm/hdbscan/tests.rs [hunk 9]
```diff
@@ -1,7 +1,296 @@
...
+            "point {i} should be in cluster C"
+        );
+    }
+}
+
+/// Uniform data with no density structure is labelled entirely noise.
+///
+/// A regular 4x4 grid has no denser sub-region that could form a cluster, so
+/// with the condensed-tree root excluded from selection (reference HDBSCAN
+/// with `allow_single_cluster = false`) every point is noise. Verified
```
Comment (michaelkirk) [inline geo/src/algorithm/hdbscan/tests.rs:276]: I don't see `allow_single_cluster` in the codebase. Is this a reference to the paper?

File: geo/src/algorithm/hdbscan/tests.rs [hunk 10]
```diff
@@ -32,6 +321,280 @@ fn test_union_find_chain() {
     assert_eq!(uf.component_label(root), 6);
 }
 
+// -- Known-answer tests (ported from petal-clustering) --------------------
```
Comment (michaelkirk) [inline geo/src/algorithm/hdbscan/tests.rs:324]: I don't know what `petal-clustering` is - can you link to it?

File: geo/src/algorithm/hdbscan/tests.rs [hunk 11]
```diff
@@ -32,6 +321,280 @@ fn test_union_find_chain() {
...
+
+/// Dataset from petal-clustering's `partial_labels` test (unsupervised).
+///
+/// Three spatial groups plus one isolated noise point:
+///   Group 1 (5 pts): indices 0..5, near (1-3, 7-9)
+///   Group 2 (4 pts): indices 5..9, near (5-6, 3-4)
+///   Group 3 (6 pts): indices 9..15, near (8-9, 1-3)
+///   Outlier:         index 15, at (7, 8)
+///
+/// With the Phase 1 correctness fixes, geo's labels and GLOSH scores match
```
Comment (michaelkirk) [inline geo/src/algorithm/hdbscan/tests.rs:334]: What are these correctness fixes about? Is this some development artifact or is this still useful for readers today?

**Review by michaelkirk — COMMENTED** [review]
I think it's actually going to be easier for me to review top down rather than bottom up.

If that's going to break everything, I guess we could leave this unmerged until I make my way down.

File: geo/src/algorithm/hdbscan/mod.rs (hunk 2 above)
Comment (urschrei) [reply geo/src/algorithm/hdbscan/mod.rs]: Seems like a good idea: https://github.com/georust/geo/pull/1594

File: geo/src/algorithm/hdbscan/mod.rs (hunk 3 above)
Comment (urschrei) [reply geo/src/algorithm/hdbscan/mod.rs]: Fixed, but it'll be fully redundant anyway if #1594 merges

File: geo/src/algorithm/hdbscan/tests.rs (hunk 8 above)
Comment (urschrei) [reply geo/src/algorithm/hdbscan/tests.rs]: Oops done. Two tests still don't use it: `test_large_n_correctness` (it would be a massive chunk of text) and `test_disconnected_spanning_tree_errors` (hopefully obvious: the wkt version would have 32 pairs with tiny magnitudes)

File: geo/src/algorithm/hdbscan/tests.rs (hunk 9 above)
Comment (urschrei) [reply geo/src/algorithm/hdbscan/tests.rs]: It's from the reference scikit-learn hdbscan library. Reworded.

File: geo/src/algorithm/hdbscan/tests.rs (hunk 10 above)
Comment (urschrei) [reply geo/src/algorithm/hdbscan/tests.rs:309]: Done.

File: geo/src/algorithm/hdbscan/tests.rs (hunk 11 above)
Comment (urschrei) [reply geo/src/algorithm/hdbscan/tests.rs]: Development artifact from earlier iteration. Replaced with something more sensible.

File: geo/src/algorithm/hdbscan/tests.rs (hunk 7 above)
Comment (urschrei) [reply geo/src/algorithm/hdbscan/tests.rs:9]: We already have the same asymmetry in `kmeans` and `dbscan`. An empty collection seems like a natural no-op to me: nothing in, nothing out. If it's non-empty but not clusterable, it's an error. The Python impl _does_ error btw. I've added a small justification to the test. WDYT?

Comment (michaelkirk) [reply geo/src/algorithm/hdbscan/tests.rs:9]: If you think it's right, we should leave it.

File: geo/src/algorithm/hdbscan/mod.rs (hunk 2 above)
Comment (michaelkirk) [reply geo/src/algorithm/hdbscan/mod.rs]: Merged! So these qualifiers and the redundant docs can go away.

**Review by michaelkirk — APPROVED** [review]

**Comment by michaelkirk** [comment]
Lol, ok I clearly don't understand how PR stacks are supposed to work. I was hoping I could merge 06 down into 05, but I guess not. Is it that we need to keep all these PR's open until the full stack is approved and then they all get merged in at once?

File: geo/src/algorithm/hdbscan/tests.rs [hunk 12]
```diff
@@ -32,6 +306,267 @@ fn test_union_find_chain() {
...
+        0.0,
+        0.0,
+        0.0,
+        0.0,
+        0.0,
+        0.0,
+        0.6,
+    ];
+    assert_eq!(result.outlier_scores.len(), expected_scores.len());
+    for (&got, &want) in result.outlier_scores.iter().zip(expected_scores.iter()) {
```
Comment (michaelkirk) [inline geo/src/algorithm/hdbscan/tests.rs:389]: minor code golf nit: approx supports slices, so you don't need to do this zipping:

```
assert_relative_eq!(result.outlier_scores.as_slice(), expected_scores.as_slice(), epsilon = 1e-9);
```

**Comment by urschrei** [comment]
> Lol, ok I clearly don't understand how PR stacks are supposed to work. I was hoping I could merge 06 down into 05, but I guess not. Is it that we need to keep all these PR's open until the full stack is approved and then they all get merged in at once?

Yep, just leave them open.
